blog: add new post about current UIWebView updates (#1096)
apache/cordova-docs: www/_posts/2020-07-18-uiwebview-warning.md

    name: Niklas Merz
    url: https://twitter.com/niklasmaerz
title:  "UPDATED: How to handle the 'Deprecated API Usage - UIWebView' warning while uploading to the App Store"
categories: howto
tags: ios
---

We recently posted [instructions](/howto/2020/03/18/wkwebviewonly.html) how to update your apps to remove all `UIWebView` references, because Apple now rejects all apps using `UIWebView`.

We are talking about this warning:

> ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs.

Some things have changed and new versions of cordova-ios have been released since the last post.

<!--more-->

## Update cordova-ios to version 6.0.0 or newer

Please update to the latest `cordova-ios` version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into `cordova-ios` and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.

> Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin your web code runs and you will loose access to web storage, such as local storage, indexed DB, etc.

Reviewer: ```suggestion
Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.
```

Reviewer: typo and grammar improvements.

---
layout: post
author:
    name: Niklas Merz
    url: https://twitter.com/niklasmaerz
title:  "UPDATED: How to handle the 'Deprecated API Usage - UIWebView' warning while uploading to the App Store"
categories: howto
tags: ios
---

We recently posted [instructions](/howto/2020/03/18/wkwebviewonly.html) how to update your apps to remove all `UIWebView` references, because Apple now rejects all apps using `UIWebView`.

We are talking about this warning:

> ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs.

Some things have changed and new versions of cordova-ios have been released since the last post.

<!--more-->

## Update cordova-ios to version 6.0.0 or newer

> Please update to the latest cordova-ios version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into cordova-iOS and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.

Reviewer: ```suggestion
Please update to the latest `cordova-ios` version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into `cordova-ios` and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.
```

Reviewer: I think we normally emphasis package names using back-ticks? Ignore if I'm wrong

> Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Using a custom scheme to serve your app content through fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file.

Reviewer 2: > Using a custom scheme to serve your app content through fixes CORS issues [...]

Is this valid English @breautek? I'd rephrase it to something like this:

> Serving your app content through a custom scheme fixes CORS issues [...]

Author: Done

> Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Using a custom scheme to serve your app content through fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file.

Reviewer: We may also want to note that changing schemes will cause clients to lose access to their current web storage, such as local storage, indexed DB, etc.

It's possible to migrate this data but Cordova doesn't currently handle this.

Author: Yeah a good idea to warn about that. Just pushed one sentence.

> Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Using a custom scheme to serve your app content through fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file.

Author: ```suggestion
Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file.
```

```xml
<preference name="scheme" value="app" />
<preference name="hostname" value="localhost" />
```

It is important to know that with the introduction of `WKURLSchemeHandler`, iOS 10 support has been dropped.

## Still getting the warning?

> If you are still getting the warning, it is most likely one or more plugins in your project are still references `UIWebView`. You will need to identify which plugins and contact those plugin's developers through their support channel (plugin's repo). They will need to fix their plugins by either remove the references or wrap them with the new flag.

Reviewer 2: ```suggestion
If you are still getting the warning, it is most likely one or more plugins in your project still references `UIWebView`. You will need to identify which plugins contain `UIWebView` code and contact those plugin's developers through their support channel (for instance their plugin's repo). They will need to fix their plugins by either removing the references or wrap them with the new flag.
```

## Using other WKWebView plugin

> The official Apache WKWebView plugin is no longer need with these cordova-ios versions since WKWebView has been integrated and UIWebView is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.

Reviewer: ```suggestion
The official Apache WKWebView plugin is no longer need with these `cordova-ios` versions since WKWebView has been integrated and UIWebView is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.
```

Reviewer: I think we normally emphasis package names using back-ticks? Ignore if I'm wrong

> The official Apache WKWebView plugin is no longer need with these cordova-ios versions since WKWebView has been integrated and UIWebView is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.

Reviewer 2: ```suggestion
The official Apache WKWebView plugin is no longer needed with these `cordova-ios` versions since `WKWebView` has been integrated and `UIWebView` is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.
```

    name: Niklas Merz
    url: https://twitter.com/niklasmaerz
title:  "UPDATED: How to handle the 'Deprecated API Usage - UIWebView' warning while uploading to the App Store"
categories: howto
tags: ios
---

We recently posted [instructions](/howto/2020/03/18/wkwebviewonly.html) how to update your apps to remove all `UIWebView` references, because Apple now rejects all apps using `UIWebView`.

We are talking about this warning:

> ITMS-90809: Deprecated API Usage - Apple will stop accepting submissions of apps that use UIWebView APIs.

Some things have changed and new versions of cordova-ios have been released since the last post.

<!--more-->

## Update cordova-ios to version 6.0.0 or newer

Please update to the latest `cordova-ios` version to get the best compatibility with recent iOS changes. As of this writing the current version is 6.1.0. Version 6.0.0 of cordova-ios moved `WKWebView` support into `cordova-ios` and removed `UIWebView` code. Due to this change, the `cordova-plugin-wkwebview-engine` plugin is obsolete and will not work with this release. If you have this plugin installed, it is safe to remove with `cordova plugin remove cordova-plugin-wkwebview-engine`.

> Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.

Reviewer 2: Is bold formatting supported? I'd argue that the sentence
> Bear in mind [...]
should be bold to catch the reader's attention because it's well hidden in the text.

> Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.

Author: ```suggestion
Additionaly, `WKURLSchemeHandler` support has been introduced with this release. Serving your app content through a custom scheme fixes CORS issues that exist because of the strict security policies that `WKWebView` has applied to the `file` scheme. You can easily configure your Cordova project to use a custom scheme by setting the preference options `scheme` and `hostname` in the `config.xml` file. **Bear in mind that running your app with a custom URL scheme changes the origin of your web code and you will lose access to web storage, such as local storage, indexed DB, etc.**
```

Good point. Bold should work.

```xml
<preference name="scheme" value="app" />
<preference name="hostname" value="localhost" />
```

It is important to know that with the introduction of `WKURLSchemeHandler`, iOS 10 support has been dropped.

## Still getting the warning?

If you are still getting the warning, it is most likely one or more plugins in your project still references `UIWebView`. You will need to identify which plugins contain `UIWebView` code and contact those plugin's developers through their support channel (for instance their plugin's repo). They will need to fix their plugins by either removing the references or wrap them with the new flag.

## Using other WKWebView plugin

The official Apache WKWebView plugin is no longer needed with these `cordova-ios` versions since `WKWebView` has been integrated and `UIWebView` is removed. There are other WKWebView plugins that can be used. Make sure they have been updated and use the latest version. If you have any issues or usage questions with their plugins, please read their docs and request help through their support channels.

## CORS issues

> If you experience any CORS issues Cordova PMC member Norman Breau published a [post on his](https://breautek.com/2020/07/14/enabling-cors/) with a good explanation of CORS.

Reviewer 2: ```suggestion
If you experience any CORS issues Cordova PMC member Norman Breau published a [post on his blog](https://breautek.com/2020/07/14/enabling-cors/) with a good explanation of CORS.
```
